File: test/support/native.c
/* Raw-stream driver. adapter.h is the only program/target-specific part. */
#include <float.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "adapter.h"

#define MAX_INSTANCES 2
#define GUARD 1234567.0f

typedef struct {
    probe state;
    float *inputs[INPUTS + 1];
    float *outputs[OUTPUTS + 1];
    int frames, next, initialized;
} stream;

static stream streams[MAX_INSTANCES];

static void fail(const char *message) {
    fprintf(stderr, "native driver: %s\n", message);
    exit(1);
}

static float *allocate(int frames) {
    float *data = malloc(((size_t)frames + 2) * sizeof(float));
    if (!data) fail("allocation failed");
    data[0] = data[frames + 1] = GUARD;
    return data + 1;
}

static float read_float(FILE *file) {
    unsigned char bytes[4];
    if (fread(bytes, 1, 4, file) != 4) fail("truncated input stream");
    uint32_t bits = (uint32_t)bytes[0] | (uint32_t)bytes[1] << 8
        | (uint32_t)bytes[2] << 16 | (uint32_t)bytes[3] << 24;
    float value;
    memcpy(&value, &bits, 4);
    return value;
}

static void write_float(FILE *file, float value) {
    uint32_t bits;
    memcpy(&bits, &value, 4);
    unsigned char bytes[4] = { bits & 255, (bits >> 8) & 255, (bits >> 16) & 255, bits >> 24 };
    if (fwrite(bytes, 1, 4, file) != 4) fail("cannot write output stream");
}

static void initialize(stream *s, int id, int frames, int poison) {
    if (s->initialized || frames < 0 || frames > 1000000 || poison < 0 || poison > 255)
        fail("invalid initialization");
    s->initialized = 1;
    s->frames = frames;
    memset(&s->state, poison, sizeof(s->state));
    probe_init(&s->state);
    char name[64];
    snprintf(name, sizeof(name), "input-%d.f32", id);
    FILE *file = fopen(name, "rb");
    if (!file) fail("cannot open input stream");
    for (int c = 0; c < INPUTS; c++) {
        s->inputs[c] = allocate(frames);
        for (int i = 0; i < frames; i++) s->inputs[c][i] = read_float(file);
    }
    if (fgetc(file) != EOF || ferror(file)) fail("wrong input byte count");
    if (fclose(file)) fail("cannot close input stream");
    for (int c = 0; c < OUTPUTS; c++) {
        s->outputs[c] = allocate(frames);
        for (int i = 0; i < frames; i++) s->outputs[c][i] = NAN;
    }
}

static void process(stream *s, int at, int frames) {
    if (at != s->next || frames < 0 || frames > s->frames - at)
        fail("invalid process range");
    const float *x[INPUTS + 1] = {0};
    float *y[OUTPUTS + 1] = {0};
    for (int c = 0; c < INPUTS; c++) x[c] = s->inputs[c] + at;
    for (int c = 0; c < OUTPUTS; c++) y[c] = s->outputs[c] + at;
    adapter_process(&s->state, x, y, frames);
    s->next += frames;
}

static void save(stream *s, int id, int pass) {
    if (s->next != s->frames || pass < 0 || pass > 1) fail("incomplete output stream");
    char name[64];
    snprintf(name, sizeof(name), "output-%d-%d.f32", id, pass);
    FILE *file = fopen(name, "wb");
    if (!file) fail("cannot open output stream");
    for (int c = 0; c < INPUTS; c++)
        if (s->inputs[c][-1] != GUARD || s->inputs[c][s->frames] != GUARD) fail("input guard changed");
    for (int c = 0; c < OUTPUTS; c++) {
        if (s->outputs[c][-1] != GUARD || s->outputs[c][s->frames] != GUARD) fail("output guard changed");
        for (int i = 0; i < s->frames; i++) {
            write_float(file, s->outputs[c][i]);
            s->outputs[c][i] = NAN;
        }
    }
    if (fclose(file)) fail("cannot close output stream");
    s->next = 0; // Rewind transport only. The next pass resets the live instance.
}

int main(int argc, char **argv) {
    if (sizeof(float) != 4 || FLT_RADIX != 2 || FLT_MANT_DIG != 24 || FLT_MAX_EXP != 128)
        fail("IEEE-754 binary32 float required");
    if (argc != 2) fail("usage: check commands.txt");
    FILE *commands = fopen(argv[1], "r");
    if (!commands) fail("cannot open commands");
    char command[16];
    while (fscanf(commands, "%15s", command) == 1) {
        int id, a, b;
        float value;
        if (fscanf(commands, "%d", &id) != 1 || id < 0 || id >= MAX_INSTANCES) fail("invalid instance");
        stream *s = &streams[id];
        if (!strcmp(command, "init")) {
            if (fscanf(commands, "%d %d", &a, &b) != 2) fail("invalid init");
            initialize(s, id, a, b);
            continue;
        }
        if (!s->initialized) fail("instance not initialized");
        if (!strcmp(command, "rate")) {
            if (fscanf(commands, "%f", &value) != 1 || !isfinite(value) || value <= 0) fail("invalid rate");
            probe_set_sample_rate(&s->state, value);
        } else if (!strcmp(command, "control")) {
            if (fscanf(commands, "%d %f", &a, &value) != 2 || a < 0 || a >= CONTROLS || !isfinite(value))
                fail("invalid control");
            adapter_control(&s->state, a, value);
        } else if (!strcmp(command, "reset")) {
            probe_reset(&s->state);
        } else if (!strcmp(command, "process")) {
            if (fscanf(commands, "%d %d", &a, &b) != 2) fail("invalid process");
            process(s, a, b);
        } else if (!strcmp(command, "save")) {
            if (fscanf(commands, "%d", &a) != 1) fail("invalid save");
            save(s, id, a);
        } else fail("unknown command");
    }
    if (ferror(commands) || fclose(commands)) fail("cannot read commands");
    for (int id = 0; id < MAX_INSTANCES; id++) {
        stream *s = &streams[id];
        if (s->next) fail("unsaved output");
        if (!s->initialized) continue;
        for (int c = 0; c < INPUTS; c++) free(s->inputs[c] - 1);
        for (int c = 0; c < OUTPUTS; c++) free(s->outputs[c] - 1);
    }
    return 0;
}

File: test/support/native.js
'use strict';

const assert = require('assert');
const fs = require('fs');
const path = require('path');
const z = require('../../src/zampogna');
const util = require('../../src/util');
const { run } = require('./process');
const { encode, decode, compare, json } = require('./streams');
const { normalize, plan, partitions } = require('./timeline');

function prepare(test, dir, optimized) {
    fs.mkdirSync(dir, { recursive: true });
    const streams = (test.streams || [test]).map(normalize);
    assert(streams.length >= 1 && streams.length <= 2, 'one or two streams required');
    const inputs = streams[0].inputs.length;
    const outputs = streams[0].expected.length;
    const controls = Object.keys(streams[0].controls);
    assert(outputs > 0, 'at least one output required');
    for (const stream of streams) {
        assert.strictEqual(stream.inputs.length, inputs, 'inconsistent input counts');
        assert.strictEqual(stream.expected.length, outputs, 'inconsistent output counts');
        assert.deepStrictEqual(Object.keys(stream.controls), controls, 'inconsistent controls');
    }
    const source = (test.prelude || '') + test.code;
    fs.writeFileSync(path.join(dir, 'source.crm'), source);
    fs.writeFileSync(path.join(dir, 'case.json'), json({ name: test.name, seed: test.seed,
        comparison: test.comparison || { mode: 'exact' }, streams }));
    for (const [name, text] of Object.entries(test.files || {})) {
        const file = path.join(dir, name);
        fs.mkdirSync(path.dirname(file), { recursive: true });
        fs.writeFileSync(file, text);
    }
    const options = { initial_block_id: 'probe', control_inputs: controls, optimizations: {
        remove_dead_graph: optimized, negative_negative: optimized,
        negative_consts: optimized, unify_consts: optimized,
    } };
    fs.writeFileSync(path.join(dir, 'compiler-options.json'), json(options));
    return { test, dir, streams, source, options, inputs, outputs, controls };
}

function generate(prepared, target) {
    const { source, options, dir } = prepared;
    const reader = util.get_filereader([dir, path.join(__dirname, '../../src')]);
    for (const file of z.compile(source, reader, { ...options, target_language: target })) {
        const dest = path.join(dir, file.path, file.name);
        fs.mkdirSync(path.dirname(dest), { recursive: true });
        fs.writeFileSync(dest, file.str);
    }
    if (target === 'bw') fs.writeFileSync(path.join(dir, 'probe/src/platform.h'), '#include <stddef.h>\n');
}

function build(prepared, target, config) {
    const { dir, inputs, outputs, controls } = prepared;
    const buildDir = path.join(dir, `${target}-${config.name}`);
    fs.mkdirSync(buildDir);
    const args = Array.from({ length: inputs }, (_, i) => `x[${i}]`)
        .concat(Array.from({ length: outputs }, (_, i) => `y[${i}]`));
    fs.writeFileSync(path.join(buildDir, 'adapter.h'), `#include "${target === 'C' ? 'probe.h' : 'probe/src/probe.h'}"
#define INPUTS ${inputs}
#define OUTPUTS ${outputs}
#define CONTROLS ${controls.length}
static void adapter_process(probe *state, const float **x, float **y, int frames) {
    (void)x; (void)y;
    probe_process(state, ${target === 'C' ? args.join(', ') : 'x, y'}, frames);
}
static void adapter_control(probe *state, int index, float value) {
    ${controls.length ? `static const int ids[] = { ${controls.map(name => 'p_' + name).join(', ')} };
    probe_set_parameter(state, ids[index], value);` : '(void)state; (void)index; (void)value;'}
}
`);
    // Keep a copy with failed builds so the executable can be rebuilt in place.
    fs.copyFileSync(path.join(__dirname, 'native.c'), path.join(buildDir, 'driver.c'));
    run(config.cc, ['-std=c99', ...config.flags,
        '-Werror=implicit-function-declaration', '-Werror=implicit-int', '-I.', '-I..',
        'driver.c', ...(target === 'bw' ? ['../probe/src/probe.c'] : []), '-lm', '-o', 'check'],
    buildDir, 'build');
    return buildDir;
}

function command(step, id, controls) {
    const number = value => Object.is(value, -0) ? '-0' : String(Math.fround(value));
    switch (step.type) {
    case 'rate': return `rate ${id} ${number(step.value)}`;
    case 'control': return `control ${id} ${controls.indexOf(step.name)} ${number(step.value)}`;
    case 'reset': return `reset ${id}`;
    case 'process': return `process ${id} ${step.at} ${step.frames}`;
    default: throw new Error('unknown step');
    }
}

function execute(prepared, buildDir, partition) {
    const { controls, test } = prepared;
    // Use two live instances even for single-stream fixtures, with different
    // storage poison and process boundaries. Explicit streams can differ too.
    const streams = prepared.streams.length === 2 ? prepared.streams : [prepared.streams[0], prepared.streams[0]];
    const dir = path.join(buildDir, partition);
    fs.mkdirSync(dir);
    const layouts = [partition, partition === 'single' ? 'irregular' : 'single'];
    const plans = streams.map((stream, id) => plan(stream, layouts[id]));
    const commands = streams.map((stream, id) => `init ${id} ${stream.frames} ${id ? 90 : 165}`);
    streams.forEach((stream, id) => {
        fs.writeFileSync(path.join(dir, `input-${id}.f32`), encode(stream.inputs, stream.frames));
        fs.writeFileSync(path.join(dir, `expected-${id}.f32`), encode(stream.expected, stream.frames));
    });
    // Interleave instances, then reset and replay without calling init again.
    for (let pass = 0; pass < 2; pass++) {
        for (let step = 0; step < Math.max(...plans.map(p => p.length)); step++)
            plans.forEach((p, id) => { if (p[step]) commands.push(command(p[step], id, controls)); });
        streams.forEach((_, id) => commands.push(`save ${id} ${pass}`));
    }
    fs.writeFileSync(path.join(dir, 'plan.json'), json({ partitions: layouts, plans }));
    fs.writeFileSync(path.join(dir, 'commands.txt'), commands.join('\n') + '\n');
    run(path.join(buildDir, 'check'), ['commands.txt'], dir, 'run', 10000);
    for (let id = 0; id < streams.length; id++) {
        for (let pass = 0; pass < 2; pass++) {
            const stream = streams[id];
            try {
                const actual = decode(fs.readFileSync(path.join(dir, `output-${id}-${pass}.f32`)),
                    stream.expected.length, stream.frames);
                compare(actual, stream.expected, stream.comparison || test.comparison);
            } catch (error) {
                throw new Error(`${partition}, instance ${id}, ${pass ? 'reset replay' : 'first run'}: ${error.message}\nArtifacts: ${dir}`);
            }
        }
    }
}

module.exports = { prepare, generate, build, execute, partitions };
